File: 01.Stack/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python
#
# [20] 有效的括号
#

# @lc code=start


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if not s or len(s) == 0:
            return True

        if len(s) % 2 == 1:
            return False

        t = []

        for c in s:
            if c == '[' or c == '{' or c == '(':
                t.append(c)
            elif c == ']':
                last = t.pop() if t else '#'
                if last != '[':
                    return False
            elif c == '}':
                last = t.pop() if t else '#'
                if last != '{':
                    return False
            elif c == ')':
                last = t.pop() if t else '#'
                if last != '(':
                    return False
            else:
                return False

        return not t

# @lc code=end

File: 01.Stack/20.有效的括号.java
/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        if (s == null || s.length() == 0) {
            return true;
        }
        if (s.length() % 2 == 1) {
            return false;
        }
        Stack<Character> t = new Stack<Character>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '{' || c == '(' || c == '[') {
                t.push(c);
            } else if (c == '}') {
                if (t.empty() || t.peek() != '{') {
                    return false;
                }
                t.pop();
            } else if (c == ']') {
                if (t.empty() || t.peek() != '[') {
                    return false;
                }
                t.pop();
            } else if (c == ')') {
                if (t.empty() || t.peek() != '(') {
                    return false;
                }
                t.pop();
            } else {
                return false;
            }
        }

        return t.empty();
    }
}
// @lc code=end

